{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "baf24aad",
   "metadata": {},
   "source": [
    "# Feature Importance\n",
    "\n",
    "In our project, we are predicting gene expression using ChIP-seq profiles. One question of interest is which epigenetics features are the most important to predict gene expression.\n",
    "\n",
    "Author: Karthik Guruvayurappan"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "3168ffb8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# for computation + data frames\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# for file operations\n",
    "import os\n",
    "\n",
    "# models\n",
    "from sklearn.tree import DecisionTreeRegressor\n",
    "from sklearn.model_selection import cross_val_score"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1a6e908",
   "metadata": {},
   "source": [
    "## Read in Epigenetic Data (from Eryk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "734705c3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# data loading code (from Eryk)\n",
    "master_folder_path = 'Multiome'\n",
    "\n",
    "dfs_epigenetics = {}\n",
    "dfs_transcriptomics = {}\n",
    "\n",
    "# Iterate through all subdirectories and files in the master folder\n",
    "for root, dirs, files in os.walk(master_folder_path):\n",
    "    # If we are at the third level of directories get the histone modification assay data\n",
    "    if len(root.split(\"/\")) == 3:\n",
    "        cell_line = root.split(\"/\")[1]\n",
    "        if cell_line not in dfs_epigenetics:\n",
    "            dfs_epigenetics[cell_line] = []\n",
    "        for file in files:\n",
    "            if file.endswith('.csv'):\n",
    "                # Construct the full path to the CSV file\n",
    "                csv_file_path = os.path.join(root, file)\n",
    "                histone_mark = csv_file_path.split(\"_\")[-1].strip(\".csv\")\n",
    "        \n",
    "                # Read the CSV file using pandas\n",
    "                df = pd.read_csv(csv_file_path)\n",
    "                df['annotation'] = df['annotation'].apply(lambda x: x.split(\" (\")[0])\n",
    "                df['annotation'] = df['annotation'] + f\"_{histone_mark}\"\n",
    "                dfs_epigenetics[cell_line].append(df)\n",
    "    # If we are at the second level of directories get RNA seq data\n",
    "    if len(root.split(\"/\")) == 2:\n",
    "        cell_line = root.split(\"/\")[1]\n",
    "        for file in files:\n",
    "            if file.endswith('.tsv'):\n",
    "                tsv_file_path = os.path.join(root, file)\n",
    "                rna_seq = pd.read_csv(tsv_file_path, sep = '\\t')\n",
    "                dfs_transcriptomics[cell_line] = rna_seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ed6584da",
   "metadata": {},
   "outputs": [],
   "source": [
    "df_aggregated = pd.concat(dfs_epigenetics['Mammary Epithelial Cell'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "39197492",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the number of peaks per gene region\n",
    "peak_count_df = df_aggregated.groupby([\"geneId\",\"annotation\"],as_index=False).agg(\n",
    "    peak_counts = ('annotation','size')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "a38b2124",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_matrix = peak_count_df.pivot_table(columns='annotation',index=['geneId'])['peak_counts']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "09589aa5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28763, 77)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "9888244b",
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_matrix.fillna(0, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "a8a11d0c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>annotation</th>\n",
       "      <th>3' UTR_H2AFZ</th>\n",
       "      <th>3' UTR_H3K27a</th>\n",
       "      <th>3' UTR_H3K27me3</th>\n",
       "      <th>3' UTR_H3K36me3</th>\n",
       "      <th>3' UTR_H3K4me1</th>\n",
       "      <th>3' UTR_H3K4me2</th>\n",
       "      <th>3' UTR_H3K4me3</th>\n",
       "      <th>3' UTR_H3K79me2</th>\n",
       "      <th>3' UTR_H3K9a</th>\n",
       "      <th>3' UTR_H3K9me3</th>\n",
       "      <th>...</th>\n",
       "      <th>Promoter_H3K27a</th>\n",
       "      <th>Promoter_H3K27me3</th>\n",
       "      <th>Promoter_H3K36me3</th>\n",
       "      <th>Promoter_H3K4me1</th>\n",
       "      <th>Promoter_H3K4me2</th>\n",
       "      <th>Promoter_H3K4me3</th>\n",
       "      <th>Promoter_H3K79me2</th>\n",
       "      <th>Promoter_H3K9a</th>\n",
       "      <th>Promoter_H3K9me3</th>\n",
       "      <th>Promoter_H4K20me1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "annotation  3' UTR_H2AFZ  3' UTR_H3K27a  3' UTR_H3K27me3  3' UTR_H3K36me3  \\\n",
       "geneId                                                                      \n",
       "1                    0.0            0.0              0.0              1.0   \n",
       "2                    0.0            0.0              0.0              0.0   \n",
       "3                    0.0            0.0              0.0              0.0   \n",
       "9                    0.0            0.0              0.0              0.0   \n",
       "10                   0.0            0.0              0.0              0.0   \n",
       "\n",
       "annotation  3' UTR_H3K4me1  3' UTR_H3K4me2  3' UTR_H3K4me3  3' UTR_H3K79me2  \\\n",
       "geneId                                                                        \n",
       "1                      0.0             0.0             0.0              0.0   \n",
       "2                      0.0             0.0             0.0              0.0   \n",
       "3                      0.0             0.0             0.0              0.0   \n",
       "9                      0.0             0.0             0.0              0.0   \n",
       "10                     0.0             0.0             0.0              0.0   \n",
       "\n",
       "annotation  3' UTR_H3K9a  3' UTR_H3K9me3  ...  Promoter_H3K27a  \\\n",
       "geneId                                    ...                    \n",
       "1                    0.0             0.0  ...              1.0   \n",
       "2                    0.0             0.0  ...              0.0   \n",
       "3                    0.0             0.0  ...              0.0   \n",
       "9                    0.0             0.0  ...              1.0   \n",
       "10                   0.0             0.0  ...              0.0   \n",
       "\n",
       "annotation  Promoter_H3K27me3  Promoter_H3K36me3  Promoter_H3K4me1  \\\n",
       "geneId                                                               \n",
       "1                         0.0                0.0               1.0   \n",
       "2                         0.0                0.0               1.0   \n",
       "3                         0.0                0.0               0.0   \n",
       "9                         0.0                1.0               0.0   \n",
       "10                        0.0                0.0               1.0   \n",
       "\n",
       "annotation  Promoter_H3K4me2  Promoter_H3K4me3  Promoter_H3K79me2  \\\n",
       "geneId                                                              \n",
       "1                        2.0               2.0                0.0   \n",
       "2                        2.0               0.0                0.0   \n",
       "3                        0.0               0.0                0.0   \n",
       "9                        1.0               1.0                0.0   \n",
       "10                       1.0               0.0                0.0   \n",
       "\n",
       "annotation  Promoter_H3K9a  Promoter_H3K9me3  Promoter_H4K20me1  \n",
       "geneId                                                           \n",
       "1                      1.0               2.0                0.0  \n",
       "2                      0.0               0.0                0.0  \n",
       "3                      0.0               0.0                0.0  \n",
       "9                      0.0               0.0                0.0  \n",
       "10                     0.0               0.0                0.0  \n",
       "\n",
       "[5 rows x 77 columns]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "278c3cbf",
   "metadata": {},
   "source": [
    "## Read in RNA-seq Data (from Eryk)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d71ca856",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq = dfs_transcriptomics['Mammary Epithelial Cell']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d59ebdf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(59526, 17)"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "76447a32",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10904</td>\n",
       "      <td>10904</td>\n",
       "      <td>93.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12954</td>\n",
       "      <td>12954</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12956</td>\n",
       "      <td>12956</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12958</td>\n",
       "      <td>12958</td>\n",
       "      <td>82.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12960</td>\n",
       "      <td>12960</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id(s)  length  effective_length  expected_count  TPM  \\\n",
       "0   10904            10904    93.0               0.0             0.0  0.0   \n",
       "1   12954            12954    94.0               0.0             0.0  0.0   \n",
       "2   12956            12956    72.0               0.0             0.0  0.0   \n",
       "3   12958            12958    82.0               0.0             0.0  0.0   \n",
       "4   12960            12960    73.0               0.0             0.0  0.0   \n",
       "\n",
       "   FPKM  posterior_mean_count  posterior_standard_deviation_of_count  pme_TPM  \\\n",
       "0   0.0                   0.0                                    0.0      0.0   \n",
       "1   0.0                   0.0                                    0.0      0.0   \n",
       "2   0.0                   0.0                                    0.0      0.0   \n",
       "3   0.0                   0.0                                    0.0      0.0   \n",
       "4   0.0                   0.0                                    0.0      0.0   \n",
       "\n",
       "   pme_FPKM  TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "0       0.0                 0.0                 0.0   \n",
       "1       0.0                 0.0                 0.0   \n",
       "2       0.0                 0.0                 0.0   \n",
       "3       0.0                 0.0                 0.0   \n",
       "4       0.0                 0.0                 0.0   \n",
       "\n",
       "   TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "0                                    0.0                  0.0   \n",
       "1                                    0.0                  0.0   \n",
       "2                                    0.0                  0.0   \n",
       "3                                    0.0                  0.0   \n",
       "4                                    0.0                  0.0   \n",
       "\n",
       "   FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \n",
       "0                  0.0                                     0.0  \n",
       "1                  0.0                                     0.0  \n",
       "2                  0.0                                     0.0  \n",
       "3                  0.0                                     0.0  \n",
       "4                  0.0                                     0.0  "
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bea45c8b",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq['gene_id_clean'] = rna_seq['gene_id'].apply(lambda x: x.split(\".\")[0])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1dc6be8b",
   "metadata": {},
   "source": [
    "## Merge datasets by gene name"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "7641f13d",
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = pd.read_table('mart_export_grch38.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "4a7be6fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>NCBI gene (formerly Entrezgene) ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>ENST00000387342.1</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>ENST00000387347.2</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>ENST00000386347.1</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0  ENSG00000210049      ENSG00000210049.1      ENST00000387314   \n",
       "1  ENSG00000211459      ENSG00000211459.2      ENST00000389680   \n",
       "2  ENSG00000210077      ENSG00000210077.1      ENST00000387342   \n",
       "3  ENSG00000210082      ENSG00000210082.2      ENST00000387347   \n",
       "4  ENSG00000209082      ENSG00000209082.1      ENST00000386347   \n",
       "\n",
       "  Transcript stable ID version Gene name  NCBI gene (formerly Entrezgene) ID  \n",
       "0            ENST00000387314.1     MT-TF                                 NaN  \n",
       "1            ENST00000389680.2   MT-RNR1                                 NaN  \n",
       "2            ENST00000387342.1     MT-TV                                 NaN  \n",
       "3            ENST00000387347.2   MT-RNR2                                 NaN  \n",
       "4            ENST00000386347.1    MT-TL1                                 NaN  "
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "6619e8c8",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gd/yfplhppn3jx0g03j2lr_yn6m0000gn/T/ipykernel_46973/2742020782.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'] = ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'].astype(int).astype(str)\n"
     ]
    }
   ],
   "source": [
    "ensembl_gene_id_map = mapping.set_index('Gene stable ID').to_dict()['Gene name']\n",
    "ensembl_transcript_id_map = mapping.set_index('Transcript stable ID version').to_dict()['Gene name']\n",
    "\n",
    "# NCBI column has NaNs which turns the integers into floats, so drop NaNs, change floats --> int --> str, and create dict\n",
    "ncbi_gene_id_map = mapping.dropna(subset=['NCBI gene (formerly Entrezgene) ID'])\n",
    "ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'] = ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'].astype(int).astype(str)\n",
    "ncbi_gene_id_map = ncbi_gene_id_map.set_index('NCBI gene (formerly Entrezgene) ID').to_dict()['Gene name']\n",
    "\n",
    "# Combine the transcript IDs, gene IDs, and NCBI IDs dictionaries for all possible mappings\n",
    "mapping_dict = ensembl_transcript_id_map | ensembl_gene_id_map | ncbi_gene_id_map "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c8eacee1",
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq['gene_name'] = rna_seq['gene_id_clean'].map(mapping_dict)\n",
    "feature_matrix['gene_name'] = feature_matrix.index.astype(str).map(mapping_dict)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "46ac2e91-6b85-4a68-8b03-37da2fdbb21a",
   "metadata": {},
   "source": [
    "## Train a model!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "b7d7a30f-9206-4119-bcbb-b799a40f396a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>annotation</th>\n",
       "      <th>3' UTR_H2AFZ</th>\n",
       "      <th>3' UTR_H3K27a</th>\n",
       "      <th>3' UTR_H3K27me3</th>\n",
       "      <th>3' UTR_H3K36me3</th>\n",
       "      <th>3' UTR_H3K4me1</th>\n",
       "      <th>3' UTR_H3K4me2</th>\n",
       "      <th>3' UTR_H3K4me3</th>\n",
       "      <th>3' UTR_H3K79me2</th>\n",
       "      <th>3' UTR_H3K9a</th>\n",
       "      <th>3' UTR_H3K9me3</th>\n",
       "      <th>...</th>\n",
       "      <th>Promoter_H3K27me3</th>\n",
       "      <th>Promoter_H3K36me3</th>\n",
       "      <th>Promoter_H3K4me1</th>\n",
       "      <th>Promoter_H3K4me2</th>\n",
       "      <th>Promoter_H3K4me3</th>\n",
       "      <th>Promoter_H3K79me2</th>\n",
       "      <th>Promoter_H3K9a</th>\n",
       "      <th>Promoter_H3K9me3</th>\n",
       "      <th>Promoter_H4K20me1</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>A1BG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>A2M</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>A2MP1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NAT1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>NAT2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 78 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "annotation  3' UTR_H2AFZ  3' UTR_H3K27a  3' UTR_H3K27me3  3' UTR_H3K36me3  \\\n",
       "geneId                                                                      \n",
       "1                    0.0            0.0              0.0              1.0   \n",
       "2                    0.0            0.0              0.0              0.0   \n",
       "3                    0.0            0.0              0.0              0.0   \n",
       "9                    0.0            0.0              0.0              0.0   \n",
       "10                   0.0            0.0              0.0              0.0   \n",
       "\n",
       "annotation  3' UTR_H3K4me1  3' UTR_H3K4me2  3' UTR_H3K4me3  3' UTR_H3K79me2  \\\n",
       "geneId                                                                        \n",
       "1                      0.0             0.0             0.0              0.0   \n",
       "2                      0.0             0.0             0.0              0.0   \n",
       "3                      0.0             0.0             0.0              0.0   \n",
       "9                      0.0             0.0             0.0              0.0   \n",
       "10                     0.0             0.0             0.0              0.0   \n",
       "\n",
       "annotation  3' UTR_H3K9a  3' UTR_H3K9me3  ...  Promoter_H3K27me3  \\\n",
       "geneId                                    ...                      \n",
       "1                    0.0             0.0  ...                0.0   \n",
       "2                    0.0             0.0  ...                0.0   \n",
       "3                    0.0             0.0  ...                0.0   \n",
       "9                    0.0             0.0  ...                0.0   \n",
       "10                   0.0             0.0  ...                0.0   \n",
       "\n",
       "annotation  Promoter_H3K36me3  Promoter_H3K4me1  Promoter_H3K4me2  \\\n",
       "geneId                                                              \n",
       "1                         0.0               1.0               2.0   \n",
       "2                         0.0               1.0               2.0   \n",
       "3                         0.0               0.0               0.0   \n",
       "9                         1.0               0.0               1.0   \n",
       "10                        0.0               1.0               1.0   \n",
       "\n",
       "annotation  Promoter_H3K4me3  Promoter_H3K79me2  Promoter_H3K9a  \\\n",
       "geneId                                                            \n",
       "1                        2.0                0.0             1.0   \n",
       "2                        0.0                0.0             0.0   \n",
       "3                        0.0                0.0             0.0   \n",
       "9                        1.0                0.0             0.0   \n",
       "10                       0.0                0.0             0.0   \n",
       "\n",
       "annotation  Promoter_H3K9me3  Promoter_H4K20me1  gene_name  \n",
       "geneId                                                      \n",
       "1                        2.0                0.0       A1BG  \n",
       "2                        0.0                0.0        A2M  \n",
       "3                        0.0                0.0      A2MP1  \n",
       "9                        0.0                0.0       NAT1  \n",
       "10                       0.0                0.0       NAT2  \n",
       "\n",
       "[5 rows x 78 columns]"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_matrix.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "6702f122-2bd0-4f80-a178-a387532596fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "      <th>gene_id_clean</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10904</td>\n",
       "      <td>10904</td>\n",
       "      <td>93.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>10904</td>\n",
       "      <td>BLCAP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12954</td>\n",
       "      <td>12954</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12954</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12956</td>\n",
       "      <td>12956</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12956</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12958</td>\n",
       "      <td>12958</td>\n",
       "      <td>82.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12958</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12960</td>\n",
       "      <td>12960</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>12960</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id(s)  length  effective_length  expected_count  TPM  \\\n",
       "0   10904            10904    93.0               0.0             0.0  0.0   \n",
       "1   12954            12954    94.0               0.0             0.0  0.0   \n",
       "2   12956            12956    72.0               0.0             0.0  0.0   \n",
       "3   12958            12958    82.0               0.0             0.0  0.0   \n",
       "4   12960            12960    73.0               0.0             0.0  0.0   \n",
       "\n",
       "   FPKM  posterior_mean_count  posterior_standard_deviation_of_count  pme_TPM  \\\n",
       "0   0.0                   0.0                                    0.0      0.0   \n",
       "1   0.0                   0.0                                    0.0      0.0   \n",
       "2   0.0                   0.0                                    0.0      0.0   \n",
       "3   0.0                   0.0                                    0.0      0.0   \n",
       "4   0.0                   0.0                                    0.0      0.0   \n",
       "\n",
       "   pme_FPKM  TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "0       0.0                 0.0                 0.0   \n",
       "1       0.0                 0.0                 0.0   \n",
       "2       0.0                 0.0                 0.0   \n",
       "3       0.0                 0.0                 0.0   \n",
       "4       0.0                 0.0                 0.0   \n",
       "\n",
       "   TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "0                                    0.0                  0.0   \n",
       "1                                    0.0                  0.0   \n",
       "2                                    0.0                  0.0   \n",
       "3                                    0.0                  0.0   \n",
       "4                                    0.0                  0.0   \n",
       "\n",
       "   FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation gene_id_clean  \\\n",
       "0                  0.0                                     0.0         10904   \n",
       "1                  0.0                                     0.0         12954   \n",
       "2                  0.0                                     0.0         12956   \n",
       "3                  0.0                                     0.0         12958   \n",
       "4                  0.0                                     0.0         12960   \n",
       "\n",
       "  gene_name  \n",
       "0     BLCAP  \n",
       "1       NaN  \n",
       "2       NaN  \n",
       "3       NaN  \n",
       "4       NaN  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "eeb5c9d2-4bdd-46b6-9a0b-8a7acf209d51",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(42026, 2)"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq = rna_seq[['gene_name', 'TPM']].dropna()\n",
    "rna_seq.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9dc41db7-462f-48c7-a2b2-159fc6fe30e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.merge(feature_matrix,rna_seq,on='gene_name').set_index('gene_name')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "1d5b2314-7db7-4374-a179-46bbde07a440",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/gd/yfplhppn3jx0g03j2lr_yn6m0000gn/T/ipykernel_46973/225978411.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  X.fillna(0, inplace = True)\n"
     ]
    }
   ],
   "source": [
    "X, y = data.iloc[:, :-1], data.iloc[:, -1]\n",
    "X.fillna(0, inplace = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "2f157107-1087-4804-9b39-4e1b261c1052",
   "metadata": {},
   "outputs": [],
   "source": [
    "y = np.log10(y+0.001)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "06034e96-52df-4847-896c-1c4489093b34",
   "metadata": {},
   "outputs": [],
   "source": [
    "model = DecisionTreeRegressor()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "7f1bcc37-9062-468a-a4cc-2cb42ea9b75e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([ 0.4912271 ,  0.28363869,  0.21742942,  0.33846877, -1.02011341])"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cross_val_score(model, X, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5c03f40d-07c2-4207-ab83-35633a6cd113",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>#sk-container-id-1 {color: black;}#sk-container-id-1 pre{padding: 0;}#sk-container-id-1 div.sk-toggleable {background-color: white;}#sk-container-id-1 label.sk-toggleable__label {cursor: pointer;display: block;width: 100%;margin-bottom: 0;padding: 0.3em;box-sizing: border-box;text-align: center;}#sk-container-id-1 label.sk-toggleable__label-arrow:before {content: \"▸\";float: left;margin-right: 0.25em;color: #696969;}#sk-container-id-1 label.sk-toggleable__label-arrow:hover:before {color: black;}#sk-container-id-1 div.sk-estimator:hover label.sk-toggleable__label-arrow:before {color: black;}#sk-container-id-1 div.sk-toggleable__content {max-height: 0;max-width: 0;overflow: hidden;text-align: left;background-color: #f0f8ff;}#sk-container-id-1 div.sk-toggleable__content pre {margin: 0.2em;color: black;border-radius: 0.25em;background-color: #f0f8ff;}#sk-container-id-1 input.sk-toggleable__control:checked~div.sk-toggleable__content {max-height: 200px;max-width: 100%;overflow: auto;}#sk-container-id-1 input.sk-toggleable__control:checked~label.sk-toggleable__label-arrow:before {content: \"▾\";}#sk-container-id-1 div.sk-estimator input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-label input.sk-toggleable__control:checked~label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 input.sk-hidden--visually {border: 0;clip: rect(1px 1px 1px 1px);clip: rect(1px, 1px, 1px, 1px);height: 1px;margin: -1px;overflow: hidden;padding: 0;position: absolute;width: 1px;}#sk-container-id-1 div.sk-estimator {font-family: monospace;background-color: #f0f8ff;border: 1px dotted black;border-radius: 0.25em;box-sizing: border-box;margin-bottom: 0.5em;}#sk-container-id-1 div.sk-estimator:hover {background-color: #d4ebff;}#sk-container-id-1 div.sk-parallel-item::after {content: \"\";width: 100%;border-bottom: 1px solid gray;flex-grow: 1;}#sk-container-id-1 div.sk-label:hover label.sk-toggleable__label {background-color: #d4ebff;}#sk-container-id-1 div.sk-serial::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: 0;}#sk-container-id-1 div.sk-serial {display: flex;flex-direction: column;align-items: center;background-color: white;padding-right: 0.2em;padding-left: 0.2em;position: relative;}#sk-container-id-1 div.sk-item {position: relative;z-index: 1;}#sk-container-id-1 div.sk-parallel {display: flex;align-items: stretch;justify-content: center;background-color: white;position: relative;}#sk-container-id-1 div.sk-item::before, #sk-container-id-1 div.sk-parallel-item::before {content: \"\";position: absolute;border-left: 1px solid gray;box-sizing: border-box;top: 0;bottom: 0;left: 50%;z-index: -1;}#sk-container-id-1 div.sk-parallel-item {display: flex;flex-direction: column;z-index: 1;position: relative;background-color: white;}#sk-container-id-1 div.sk-parallel-item:first-child::after {align-self: flex-end;width: 50%;}#sk-container-id-1 div.sk-parallel-item:last-child::after {align-self: flex-start;width: 50%;}#sk-container-id-1 div.sk-parallel-item:only-child::after {width: 0;}#sk-container-id-1 div.sk-dashed-wrapped {border: 1px dashed gray;margin: 0 0.4em 0.5em 0.4em;box-sizing: border-box;padding-bottom: 0.4em;background-color: white;}#sk-container-id-1 div.sk-label label {font-family: monospace;font-weight: bold;display: inline-block;line-height: 1.2em;}#sk-container-id-1 div.sk-label-container {text-align: center;}#sk-container-id-1 div.sk-container {/* jupyter's `normalize.less` sets `[hidden] { display: none; }` but bootstrap.min.css set `[hidden] { display: none !important; }` so we also need the `!important` here to be able to override the default hidden behavior on the sphinx rendered scikit-learn.org. See: https://github.com/scikit-learn/scikit-learn/issues/21755 */display: inline-block !important;position: relative;}#sk-container-id-1 div.sk-text-repr-fallback {display: none;}</style><div id=\"sk-container-id-1\" class=\"sk-top-container\"><div class=\"sk-text-repr-fallback\"><pre>DecisionTreeRegressor()</pre><b>In a Jupyter environment, please rerun this cell to show the HTML representation or trust the notebook. <br />On GitHub, the HTML representation is unable to render, please try loading this page with nbviewer.org.</b></div><div class=\"sk-container\" hidden><div class=\"sk-item\"><div class=\"sk-estimator sk-toggleable\"><input class=\"sk-toggleable__control sk-hidden--visually\" id=\"sk-estimator-id-1\" type=\"checkbox\" checked><label for=\"sk-estimator-id-1\" class=\"sk-toggleable__label sk-toggleable__label-arrow\">DecisionTreeRegressor</label><div class=\"sk-toggleable__content\"><pre>DecisionTreeRegressor()</pre></div></div></div></div></div>"
      ],
      "text/plain": [
       "DecisionTreeRegressor()"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.fit(X, y)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "b7e80da3-e4f8-4876-87fc-4cda97112159",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "77"
      ]
     },
     "execution_count": 34,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.feature_names_in_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "cf7c1114-c400-4746-ac4d-2a2d0ac1133b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "77"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(model.feature_importances_)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "7dd0cd4a-5e06-4a3d-877b-10eacb1cc535",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>feature</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>74</th>\n",
       "      <td>Promoter_H3K9a</td>\n",
       "      <td>0.51982</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>73</th>\n",
       "      <td>Promoter_H3K79me2</td>\n",
       "      <td>0.094367</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>72</th>\n",
       "      <td>Promoter_H3K4me3</td>\n",
       "      <td>0.035603</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>70</th>\n",
       "      <td>Promoter_H3K4me1</td>\n",
       "      <td>0.028647</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>69</th>\n",
       "      <td>Promoter_H3K36me3</td>\n",
       "      <td>0.026318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>66</th>\n",
       "      <td>Promoter_H2AFZ</td>\n",
       "      <td>0.0168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>58</th>\n",
       "      <td>Intron_H3K36me3</td>\n",
       "      <td>0.016694</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>71</th>\n",
       "      <td>Promoter_H3K4me2</td>\n",
       "      <td>0.016072</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>3' UTR_H3K36me3</td>\n",
       "      <td>0.015268</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>62</th>\n",
       "      <td>Intron_H3K79me2</td>\n",
       "      <td>0.014401</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "              feature     score\n",
       "74     Promoter_H3K9a   0.51982\n",
       "73  Promoter_H3K79me2  0.094367\n",
       "72   Promoter_H3K4me3  0.035603\n",
       "70   Promoter_H3K4me1  0.028647\n",
       "69  Promoter_H3K36me3  0.026318\n",
       "66     Promoter_H2AFZ    0.0168\n",
       "58    Intron_H3K36me3  0.016694\n",
       "71   Promoter_H3K4me2  0.016072\n",
       "3     3' UTR_H3K36me3  0.015268\n",
       "62    Intron_H3K79me2  0.014401"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "features_df = pd.DataFrame([model.feature_names_in_, model.feature_importances_]).T\n",
    "features_df.columns = ['feature', 'score']\n",
    "features_df.sort_values(by = 'score', ascending = False).head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b0a29689-c07c-48b5-8633-4d384e930557",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
