{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import Epigenetics and RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_folder_path = 'Multiome'\n",
    "\n",
    "dfs_epigenetics = {}\n",
    "dfs_transcriptomics = {}\n",
    "\n",
    "# Iterate through all subdirectories and files in the master folder\n",
    "for root, dirs, files in os.walk(master_folder_path):\n",
    "    # If we are at the third level of directories get the histone modification assay data\n",
    "    if len(root.split(\"/\")) == 3:\n",
    "        cell_line = root.split(\"/\")[1]\n",
    "        if cell_line not in dfs_epigenetics:\n",
    "            dfs_epigenetics[cell_line] = []\n",
    "        for file in files:\n",
    "            if file.endswith('.csv'):\n",
    "                # Construct the full path to the CSV file\n",
    "                csv_file_path = os.path.join(root, file)\n",
    "                histone_mark = csv_file_path.split(\"_\")[-1].strip(\".csv\")\n",
    "        \n",
    "                # Read the CSV file using pandas\n",
    "                df = pd.read_csv(csv_file_path)\n",
    "                df['annotation'] = df['annotation'].apply(lambda x: x.split(\" (\")[0])\n",
    "                df['annotation'] = df['annotation'] + f\"_{histone_mark}\"\n",
    "                dfs_epigenetics[cell_line].append(df)\n",
    "    # If we are at the second level of directories get RNA seq data\n",
    "    if len(root.split(\"/\")) == 2:\n",
    "        cell_line = root.split(\"/\")[1]\n",
    "        for file in files:\n",
    "            if file.endswith('.tsv'):\n",
    "                tsv_file_path = os.path.join(root, file)\n",
    "                rna_seq = pd.read_csv(tsv_file_path, sep = '\\t')\n",
    "                dfs_transcriptomics[cell_line] = rna_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lets use Mammary Epithelial Cell as an example for now and create the feature matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take this as an example for now\n",
    "df_aggregated = pd.concat(dfs_epigenetics['Mammary Epithelial Cell'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>geneChr</th>\n",
       "      <th>geneStart</th>\n",
       "      <th>geneEnd</th>\n",
       "      <th>geneLength</th>\n",
       "      <th>geneStrand</th>\n",
       "      <th>geneId</th>\n",
       "      <th>transcriptId</th>\n",
       "      <th>distanceToTSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100035284</td>\n",
       "      <td>100036150</td>\n",
       "      <td>867</td>\n",
       "      <td>*</td>\n",
       "      <td>3' UTR_H3K4me2</td>\n",
       "      <td>1</td>\n",
       "      <td>100038095</td>\n",
       "      <td>100083377</td>\n",
       "      <td>45283</td>\n",
       "      <td>1</td>\n",
       "      <td>64645</td>\n",
       "      <td>ENST00000370152.8</td>\n",
       "      <td>-1945</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100036465</td>\n",
       "      <td>100038061</td>\n",
       "      <td>1597</td>\n",
       "      <td>*</td>\n",
       "      <td>Promoter_H3K4me2</td>\n",
       "      <td>1</td>\n",
       "      <td>100038095</td>\n",
       "      <td>100083377</td>\n",
       "      <td>45283</td>\n",
       "      <td>1</td>\n",
       "      <td>64645</td>\n",
       "      <td>ENST00000370152.8</td>\n",
       "      <td>-34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100038100</td>\n",
       "      <td>100040871</td>\n",
       "      <td>2772</td>\n",
       "      <td>*</td>\n",
       "      <td>Promoter_H3K4me2</td>\n",
       "      <td>1</td>\n",
       "      <td>100038095</td>\n",
       "      <td>100083377</td>\n",
       "      <td>45283</td>\n",
       "      <td>1</td>\n",
       "      <td>64645</td>\n",
       "      <td>ENST00000370152.8</td>\n",
       "      <td>5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100042106</td>\n",
       "      <td>100042601</td>\n",
       "      <td>496</td>\n",
       "      <td>*</td>\n",
       "      <td>Intron_H3K4me2</td>\n",
       "      <td>1</td>\n",
       "      <td>100038095</td>\n",
       "      <td>100083377</td>\n",
       "      <td>45283</td>\n",
       "      <td>1</td>\n",
       "      <td>64645</td>\n",
       "      <td>ENST00000370152.8</td>\n",
       "      <td>4011</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100042648</td>\n",
       "      <td>100042938</td>\n",
       "      <td>291</td>\n",
       "      <td>*</td>\n",
       "      <td>Intron_H3K4me2</td>\n",
       "      <td>1</td>\n",
       "      <td>100038095</td>\n",
       "      <td>100083377</td>\n",
       "      <td>45283</td>\n",
       "      <td>1</td>\n",
       "      <td>64645</td>\n",
       "      <td>ENST00000370152.8</td>\n",
       "      <td>4553</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0 seqnames      start        end  width strand        annotation  \\\n",
       "0           1     chr1  100035284  100036150    867      *    3' UTR_H3K4me2   \n",
       "1           2     chr1  100036465  100038061   1597      *  Promoter_H3K4me2   \n",
       "2           3     chr1  100038100  100040871   2772      *  Promoter_H3K4me2   \n",
       "3           4     chr1  100042106  100042601    496      *    Intron_H3K4me2   \n",
       "4           5     chr1  100042648  100042938    291      *    Intron_H3K4me2   \n",
       "\n",
       "   geneChr  geneStart    geneEnd  geneLength  geneStrand  geneId  \\\n",
       "0        1  100038095  100083377       45283           1   64645   \n",
       "1        1  100038095  100083377       45283           1   64645   \n",
       "2        1  100038095  100083377       45283           1   64645   \n",
       "3        1  100038095  100083377       45283           1   64645   \n",
       "4        1  100038095  100083377       45283           1   64645   \n",
       "\n",
       "        transcriptId  distanceToTSS  \n",
       "0  ENST00000370152.8          -1945  \n",
       "1  ENST00000370152.8            -34  \n",
       "2  ENST00000370152.8              5  \n",
       "3  ENST00000370152.8           4011  \n",
       "4  ENST00000370152.8           4553  "
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_aggregated.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the number of peaks per gene region\n",
    "peak_count_df = df_aggregated.groupby([\"geneId\",\"annotation\"],as_index=False).agg(\n",
    "    peak_counts = ('annotation','size')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_matrix = peak_count_df.pivot_table(columns='annotation',index=['geneId'])['peak_counts']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(28763, 77)"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>annotation</th>\n",
       "      <th>3' UTR_H2AFZ</th>\n",
       "      <th>3' UTR_H3K27a</th>\n",
       "      <th>3' UTR_H3K27me3</th>\n",
       "      <th>3' UTR_H3K36me3</th>\n",
       "      <th>3' UTR_H3K4me1</th>\n",
       "      <th>3' UTR_H3K4me2</th>\n",
       "      <th>3' UTR_H3K4me3</th>\n",
       "      <th>3' UTR_H3K79me2</th>\n",
       "      <th>3' UTR_H3K9a</th>\n",
       "      <th>3' UTR_H3K9me3</th>\n",
       "      <th>...</th>\n",
       "      <th>Promoter_H3K27a</th>\n",
       "      <th>Promoter_H3K27me3</th>\n",
       "      <th>Promoter_H3K36me3</th>\n",
       "      <th>Promoter_H3K4me1</th>\n",
       "      <th>Promoter_H3K4me2</th>\n",
       "      <th>Promoter_H3K4me3</th>\n",
       "      <th>Promoter_H3K79me2</th>\n",
       "      <th>Promoter_H3K9a</th>\n",
       "      <th>Promoter_H3K9me3</th>\n",
       "      <th>Promoter_H4K20me1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 77 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "annotation  3' UTR_H2AFZ  3' UTR_H3K27a  3' UTR_H3K27me3  3' UTR_H3K36me3  \\\n",
       "geneId                                                                      \n",
       "1                    NaN            NaN              NaN              1.0   \n",
       "2                    NaN            NaN              NaN              NaN   \n",
       "3                    NaN            NaN              NaN              NaN   \n",
       "9                    NaN            NaN              NaN              NaN   \n",
       "10                   NaN            NaN              NaN              NaN   \n",
       "\n",
       "annotation  3' UTR_H3K4me1  3' UTR_H3K4me2  3' UTR_H3K4me3  3' UTR_H3K79me2  \\\n",
       "geneId                                                                        \n",
       "1                      NaN             NaN             NaN              NaN   \n",
       "2                      NaN             NaN             NaN              NaN   \n",
       "3                      NaN             NaN             NaN              NaN   \n",
       "9                      NaN             NaN             NaN              NaN   \n",
       "10                     NaN             NaN             NaN              NaN   \n",
       "\n",
       "annotation  3' UTR_H3K9a  3' UTR_H3K9me3  ...  Promoter_H3K27a  \\\n",
       "geneId                                    ...                    \n",
       "1                    NaN             NaN  ...              1.0   \n",
       "2                    NaN             NaN  ...              NaN   \n",
       "3                    NaN             NaN  ...              NaN   \n",
       "9                    NaN             NaN  ...              1.0   \n",
       "10                   NaN             NaN  ...              NaN   \n",
       "\n",
       "annotation  Promoter_H3K27me3  Promoter_H3K36me3  Promoter_H3K4me1  \\\n",
       "geneId                                                               \n",
       "1                         NaN                NaN               1.0   \n",
       "2                         NaN                NaN               1.0   \n",
       "3                         NaN                NaN               NaN   \n",
       "9                         NaN                1.0               NaN   \n",
       "10                        NaN                NaN               1.0   \n",
       "\n",
       "annotation  Promoter_H3K4me2  Promoter_H3K4me3  Promoter_H3K79me2  \\\n",
       "geneId                                                              \n",
       "1                        2.0               2.0                NaN   \n",
       "2                        2.0               NaN                NaN   \n",
       "3                        NaN               NaN                NaN   \n",
       "9                        1.0               1.0                NaN   \n",
       "10                       1.0               NaN                NaN   \n",
       "\n",
       "annotation  Promoter_H3K9a  Promoter_H3K9me3  Promoter_H4K20me1  \n",
       "geneId                                                           \n",
       "1                      1.0               2.0                NaN  \n",
       "2                      NaN               NaN                NaN  \n",
       "3                      NaN               NaN                NaN  \n",
       "9                      NaN               NaN                NaN  \n",
       "10                     NaN               NaN                NaN  \n",
       "\n",
       "[5 rows x 77 columns]"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq = dfs_transcriptomics['Mammary Epithelial Cell']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10904</td>\n",
       "      <td>10904</td>\n",
       "      <td>93.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>12954</td>\n",
       "      <td>12954</td>\n",
       "      <td>94.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>12956</td>\n",
       "      <td>12956</td>\n",
       "      <td>72.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>12958</td>\n",
       "      <td>12958</td>\n",
       "      <td>82.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>12960</td>\n",
       "      <td>12960</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gene_id transcript_id(s)  length  effective_length  expected_count  TPM  \\\n",
       "0   10904            10904    93.0               0.0             0.0  0.0   \n",
       "1   12954            12954    94.0               0.0             0.0  0.0   \n",
       "2   12956            12956    72.0               0.0             0.0  0.0   \n",
       "3   12958            12958    82.0               0.0             0.0  0.0   \n",
       "4   12960            12960    73.0               0.0             0.0  0.0   \n",
       "\n",
       "   FPKM  posterior_mean_count  posterior_standard_deviation_of_count  pme_TPM  \\\n",
       "0   0.0                   0.0                                    0.0      0.0   \n",
       "1   0.0                   0.0                                    0.0      0.0   \n",
       "2   0.0                   0.0                                    0.0      0.0   \n",
       "3   0.0                   0.0                                    0.0      0.0   \n",
       "4   0.0                   0.0                                    0.0      0.0   \n",
       "\n",
       "   pme_FPKM  TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "0       0.0                 0.0                 0.0   \n",
       "1       0.0                 0.0                 0.0   \n",
       "2       0.0                 0.0                 0.0   \n",
       "3       0.0                 0.0                 0.0   \n",
       "4       0.0                 0.0                 0.0   \n",
       "\n",
       "   TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "0                                    0.0                  0.0   \n",
       "1                                    0.0                  0.0   \n",
       "2                                    0.0                  0.0   \n",
       "3                                    0.0                  0.0   \n",
       "4                                    0.0                  0.0   \n",
       "\n",
       "   FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \n",
       "0                  0.0                                     0.0  \n",
       "1                  0.0                                     0.0  \n",
       "2                  0.0                                     0.0  \n",
       "3                  0.0                                     0.0  \n",
       "4                  0.0                                     0.0  "
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lets harmonizome the gene identifiers in the epigenetics data and RNA-seq data to gene names "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = pd.read_table('mart_export_grch38.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>NCBI gene (formerly Entrezgene) ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>ENST00000387342.1</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>ENST00000387347.2</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>ENST00000386347.1</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0  ENSG00000210049      ENSG00000210049.1      ENST00000387314   \n",
       "1  ENSG00000211459      ENSG00000211459.2      ENST00000389680   \n",
       "2  ENSG00000210077      ENSG00000210077.1      ENST00000387342   \n",
       "3  ENSG00000210082      ENSG00000210082.2      ENST00000387347   \n",
       "4  ENSG00000209082      ENSG00000209082.1      ENST00000386347   \n",
       "\n",
       "  Transcript stable ID version Gene name  NCBI gene (formerly Entrezgene) ID  \n",
       "0            ENST00000387314.1     MT-TF                                 NaN  \n",
       "1            ENST00000389680.2   MT-RNR1                                 NaN  \n",
       "2            ENST00000387342.1     MT-TV                                 NaN  \n",
       "3            ENST00000387347.2   MT-RNR2                                 NaN  \n",
       "4            ENST00000386347.1    MT-TL1                                 NaN  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/p4/gdh28gk568sbf143d4p59vdw0000gq/T/ipykernel_13939/757272775.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'] = ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'].astype(int).astype(str)\n"
     ]
    }
   ],
   "source": [
    "ensembl_gene_id_map = mapping.set_index('Gene stable ID version').to_dict()['Gene name']\n",
    "ensembl_transcript_id_map = mapping.set_index('Transcript stable ID version').to_dict()['Gene name']\n",
    "\n",
    "# NCBI column has NaNs which turns the integers into floats, so drop NaNs, change floats --> int --> str, and create dict\n",
    "ncbi_gene_id_map = mapping.dropna(subset='NCBI gene (formerly Entrezgene) ID')\n",
    "ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'] = ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'].astype(int).astype(str)\n",
    "ncbi_gene_id_map = ncbi_gene_id_map.set_index('NCBI gene (formerly Entrezgene) ID').to_dict()['Gene name']\n",
    "\n",
    "# Combine the transcript IDs, gene IDs, and NCBI IDs dictionaries for all possible mappings\n",
    "mapping_dict = ensembl_transcript_id_map | ensembl_gene_id_map | ncbi_gene_id_map "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq['gene_name'] = rna_seq['gene_id'].map(mapping_dict)\n",
    "feature_matrix['gene_name'] = feature_matrix.index.astype(str).map(mapping_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq.dropna(subset='gene_name', inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The below seems like a lot of genes aren't mapped between the epigenetics and RNA-seq datasets... this limits our training data quite a bit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3' UTR_H2AFZ</th>\n",
       "      <th>3' UTR_H3K27a</th>\n",
       "      <th>3' UTR_H3K27me3</th>\n",
       "      <th>3' UTR_H3K36me3</th>\n",
       "      <th>3' UTR_H3K4me1</th>\n",
       "      <th>3' UTR_H3K4me2</th>\n",
       "      <th>3' UTR_H3K4me3</th>\n",
       "      <th>3' UTR_H3K79me2</th>\n",
       "      <th>3' UTR_H3K9a</th>\n",
       "      <th>3' UTR_H3K9me3</th>\n",
       "      <th>...</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>236.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>8.73</td>\n",
       "      <td>9.28</td>\n",
       "      <td>7.589790</td>\n",
       "      <td>9.902190</td>\n",
       "      <td>0.045273</td>\n",
       "      <td>8.060280</td>\n",
       "      <td>10.516200</td>\n",
       "      <td>0.045239</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.37</td>\n",
       "      <td>0.39</td>\n",
       "      <td>0.110687</td>\n",
       "      <td>0.676154</td>\n",
       "      <td>0.274713</td>\n",
       "      <td>0.117439</td>\n",
       "      <td>0.718343</td>\n",
       "      <td>0.274743</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>11444.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>99.75</td>\n",
       "      <td>105.99</td>\n",
       "      <td>96.428500</td>\n",
       "      <td>103.203000</td>\n",
       "      <td>0.011759</td>\n",
       "      <td>102.481000</td>\n",
       "      <td>109.678000</td>\n",
       "      <td>0.011786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>3032.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>27.99</td>\n",
       "      <td>29.74</td>\n",
       "      <td>26.401500</td>\n",
       "      <td>29.539100</td>\n",
       "      <td>0.019641</td>\n",
       "      <td>28.076200</td>\n",
       "      <td>31.414700</td>\n",
       "      <td>0.019630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.17</td>\n",
       "      <td>0.016836</td>\n",
       "      <td>0.352125</td>\n",
       "      <td>0.389764</td>\n",
       "      <td>0.017881</td>\n",
       "      <td>0.374190</td>\n",
       "      <td>0.389839</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3997</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.03</td>\n",
       "      <td>0.04</td>\n",
       "      <td>0.000592</td>\n",
       "      <td>0.078898</td>\n",
       "      <td>0.474567</td>\n",
       "      <td>0.000847</td>\n",
       "      <td>0.084061</td>\n",
       "      <td>0.474802</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3998</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.003261</td>\n",
       "      <td>0.359758</td>\n",
       "      <td>0.480782</td>\n",
       "      <td>0.001904</td>\n",
       "      <td>0.380769</td>\n",
       "      <td>0.480732</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3999</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.09</td>\n",
       "      <td>0.002408</td>\n",
       "      <td>0.211425</td>\n",
       "      <td>0.471208</td>\n",
       "      <td>0.002557</td>\n",
       "      <td>0.224624</td>\n",
       "      <td>0.470914</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4000</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>4.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.63</td>\n",
       "      <td>0.67</td>\n",
       "      <td>0.177834</td>\n",
       "      <td>1.164040</td>\n",
       "      <td>0.286296</td>\n",
       "      <td>0.188051</td>\n",
       "      <td>1.236190</td>\n",
       "      <td>0.286235</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4001</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.002150</td>\n",
       "      <td>0.348809</td>\n",
       "      <td>0.473038</td>\n",
       "      <td>0.002923</td>\n",
       "      <td>0.371405</td>\n",
       "      <td>0.473064</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4002 rows × 95 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      3' UTR_H2AFZ  3' UTR_H3K27a  3' UTR_H3K27me3  3' UTR_H3K36me3  \\\n",
       "0              NaN            NaN              NaN              NaN   \n",
       "1              NaN            NaN              NaN              NaN   \n",
       "2              1.0            1.0              NaN              2.0   \n",
       "3              NaN            NaN              NaN              3.0   \n",
       "4              NaN            NaN              NaN              NaN   \n",
       "...            ...            ...              ...              ...   \n",
       "3997           NaN            NaN              NaN              NaN   \n",
       "3998           NaN            NaN              NaN              NaN   \n",
       "3999           NaN            NaN              NaN              NaN   \n",
       "4000           NaN            NaN              NaN              NaN   \n",
       "4001           NaN            NaN              NaN              NaN   \n",
       "\n",
       "      3' UTR_H3K4me1  3' UTR_H3K4me2  3' UTR_H3K4me3  3' UTR_H3K79me2  \\\n",
       "0                NaN             NaN             NaN              NaN   \n",
       "1                NaN             NaN             NaN              NaN   \n",
       "2                3.0             2.0             NaN              1.0   \n",
       "3                NaN             NaN             NaN              NaN   \n",
       "4                NaN             NaN             NaN              NaN   \n",
       "...              ...             ...             ...              ...   \n",
       "3997             NaN             NaN             NaN              NaN   \n",
       "3998             NaN             NaN             NaN              NaN   \n",
       "3999             NaN             NaN             NaN              NaN   \n",
       "4000             NaN             NaN             NaN              NaN   \n",
       "4001             NaN             NaN             NaN              NaN   \n",
       "\n",
       "      3' UTR_H3K9a  3' UTR_H3K9me3  ...  posterior_mean_count  \\\n",
       "0              NaN             NaN  ...                 236.0   \n",
       "1              NaN             NaN  ...                   0.0   \n",
       "2              NaN             NaN  ...               11444.0   \n",
       "3              NaN             NaN  ...                3032.0   \n",
       "4              NaN             NaN  ...                   0.0   \n",
       "...            ...             ...  ...                   ...   \n",
       "3997           NaN             NaN  ...                   0.0   \n",
       "3998           NaN             NaN  ...                   0.0   \n",
       "3999           NaN             NaN  ...                   1.0   \n",
       "4000           NaN             NaN  ...                   4.0   \n",
       "4001           NaN             NaN  ...                   0.0   \n",
       "\n",
       "      posterior_standard_deviation_of_count  pme_TPM  pme_FPKM  \\\n",
       "0                                       0.0     8.73      9.28   \n",
       "1                                       0.0     0.37      0.39   \n",
       "2                                       0.0    99.75    105.99   \n",
       "3                                       0.0    27.99     29.74   \n",
       "4                                       0.0     0.16      0.17   \n",
       "...                                     ...      ...       ...   \n",
       "3997                                    0.0     0.03      0.04   \n",
       "3998                                    0.0     0.15      0.16   \n",
       "3999                                    0.0     0.09      0.09   \n",
       "4000                                    0.0     0.63      0.67   \n",
       "4001                                    0.0     0.15      0.16   \n",
       "\n",
       "      TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "0               7.589790            9.902190   \n",
       "1               0.110687            0.676154   \n",
       "2              96.428500          103.203000   \n",
       "3              26.401500           29.539100   \n",
       "4               0.016836            0.352125   \n",
       "...                  ...                 ...   \n",
       "3997            0.000592            0.078898   \n",
       "3998            0.003261            0.359758   \n",
       "3999            0.002408            0.211425   \n",
       "4000            0.177834            1.164040   \n",
       "4001            0.002150            0.348809   \n",
       "\n",
       "      TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "0                                  0.045273             8.060280   \n",
       "1                                  0.274713             0.117439   \n",
       "2                                  0.011759           102.481000   \n",
       "3                                  0.019641            28.076200   \n",
       "4                                  0.389764             0.017881   \n",
       "...                                     ...                  ...   \n",
       "3997                               0.474567             0.000847   \n",
       "3998                               0.480782             0.001904   \n",
       "3999                               0.471208             0.002557   \n",
       "4000                               0.286296             0.188051   \n",
       "4001                               0.473038             0.002923   \n",
       "\n",
       "      FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \n",
       "0               10.516200                                0.045239  \n",
       "1                0.718343                                0.274743  \n",
       "2              109.678000                                0.011786  \n",
       "3               31.414700                                0.019630  \n",
       "4                0.374190                                0.389839  \n",
       "...                   ...                                     ...  \n",
       "3997             0.084061                                0.474802  \n",
       "3998             0.380769                                0.480732  \n",
       "3999             0.224624                                0.470914  \n",
       "4000             1.236190                                0.286235  \n",
       "4001             0.371405                                0.473064  \n",
       "\n",
       "[4002 rows x 95 columns]"
      ]
     },
     "execution_count": 48,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(feature_matrix,rna_seq,on='gene_name')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biohack",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
