{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import os"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import Epigenetics and RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "master_folder_path = 'Multiome'\n",
    "\n",
    "dfs_epigenetics = {}\n",
    "dfs_transcriptomics = {}\n",
    "\n",
    "# Iterate through all subdirectories and files in the master folder\n",
    "for root, dirs, files in os.walk(master_folder_path):\n",
    "    # If we are at the third level of directories get the histone modification assay data\n",
    "    if len(root.split(\"/\")) == 3:\n",
    "        cell_line = root.split(\"/\")[1]\n",
    "        dfs_epigenetics[cell_line] = []\n",
    "        for file in files:\n",
    "            if file.endswith('.csv'):\n",
    "                # Construct the full path to the CSV file\n",
    "                csv_file_path = os.path.join(root, file)\n",
    "                histone_mark = csv_file_path.split(\"_\")[-1].strip(\".csv\")\n",
    "        \n",
    "                # Read the CSV file using pandas\n",
    "                df = pd.read_csv(csv_file_path)\n",
    "                df['annotation'] = df['annotation'].apply(lambda x: x.split(\" (\")[0])\n",
    "                df['annotation'] = df['annotation'] + f\"_{histone_mark}\"\n",
    "                dfs_epigenetics[cell_line].append(df)\n",
    "    # If we are at the second level of directories get RNA seq data\n",
    "    if len(root.split(\"/\")) == 2:\n",
    "        cell_line = root.split(\"/\")[1]\n",
    "        for file in files:\n",
    "            if file.endswith('.tsv'):\n",
    "                tsv_file_path = os.path.join(root, file)\n",
    "                rna_seq = pd.read_csv(tsv_file_path, sep = '\\t')\n",
    "                dfs_transcriptomics[cell_line] = rna_seq"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lets use Mammary Epithelial Cell as an example for now and create the feature matrix"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Take this as an example for now\n",
    "df_aggregated = pd.concat(dfs_epigenetics['Mammary Epithelial Cell'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>seqnames</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>width</th>\n",
       "      <th>strand</th>\n",
       "      <th>annotation</th>\n",
       "      <th>geneChr</th>\n",
       "      <th>geneStart</th>\n",
       "      <th>geneEnd</th>\n",
       "      <th>geneLength</th>\n",
       "      <th>geneStrand</th>\n",
       "      <th>geneId</th>\n",
       "      <th>transcriptId</th>\n",
       "      <th>distanceToTSS</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100000388</td>\n",
       "      <td>100000681</td>\n",
       "      <td>294</td>\n",
       "      <td>*</td>\n",
       "      <td>Exon_H3K36me3</td>\n",
       "      <td>1</td>\n",
       "      <td>99993546</td>\n",
       "      <td>100023411</td>\n",
       "      <td>29866</td>\n",
       "      <td>1</td>\n",
       "      <td>23443</td>\n",
       "      <td>ENST00000639148.1</td>\n",
       "      <td>6842</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100008333</td>\n",
       "      <td>100009036</td>\n",
       "      <td>704</td>\n",
       "      <td>*</td>\n",
       "      <td>Exon_H3K36me3</td>\n",
       "      <td>1</td>\n",
       "      <td>99993546</td>\n",
       "      <td>100023411</td>\n",
       "      <td>29866</td>\n",
       "      <td>1</td>\n",
       "      <td>23443</td>\n",
       "      <td>ENST00000639148.1</td>\n",
       "      <td>14787</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>3</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100013273</td>\n",
       "      <td>100013497</td>\n",
       "      <td>225</td>\n",
       "      <td>*</td>\n",
       "      <td>Intron_H3K36me3</td>\n",
       "      <td>1</td>\n",
       "      <td>99993546</td>\n",
       "      <td>100023411</td>\n",
       "      <td>29866</td>\n",
       "      <td>1</td>\n",
       "      <td>23443</td>\n",
       "      <td>ENST00000639148.1</td>\n",
       "      <td>19727</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>4</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100013752</td>\n",
       "      <td>100014072</td>\n",
       "      <td>321</td>\n",
       "      <td>*</td>\n",
       "      <td>Intron_H3K36me3</td>\n",
       "      <td>1</td>\n",
       "      <td>99993546</td>\n",
       "      <td>100023411</td>\n",
       "      <td>29866</td>\n",
       "      <td>1</td>\n",
       "      <td>23443</td>\n",
       "      <td>ENST00000639148.1</td>\n",
       "      <td>20206</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>5</td>\n",
       "      <td>chr1</td>\n",
       "      <td>100015168</td>\n",
       "      <td>100015392</td>\n",
       "      <td>225</td>\n",
       "      <td>*</td>\n",
       "      <td>3' UTR_H3K36me3</td>\n",
       "      <td>1</td>\n",
       "      <td>99993546</td>\n",
       "      <td>100023411</td>\n",
       "      <td>29866</td>\n",
       "      <td>1</td>\n",
       "      <td>23443</td>\n",
       "      <td>ENST00000639148.1</td>\n",
       "      <td>21622</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Unnamed: 0 seqnames      start        end  width strand       annotation  \\\n",
       "0           1     chr1  100000388  100000681    294      *    Exon_H3K36me3   \n",
       "1           2     chr1  100008333  100009036    704      *    Exon_H3K36me3   \n",
       "2           3     chr1  100013273  100013497    225      *  Intron_H3K36me3   \n",
       "3           4     chr1  100013752  100014072    321      *  Intron_H3K36me3   \n",
       "4           5     chr1  100015168  100015392    225      *  3' UTR_H3K36me3   \n",
       "\n",
       "   geneChr  geneStart    geneEnd  geneLength  geneStrand  geneId  \\\n",
       "0        1   99993546  100023411       29866           1   23443   \n",
       "1        1   99993546  100023411       29866           1   23443   \n",
       "2        1   99993546  100023411       29866           1   23443   \n",
       "3        1   99993546  100023411       29866           1   23443   \n",
       "4        1   99993546  100023411       29866           1   23443   \n",
       "\n",
       "        transcriptId  distanceToTSS  \n",
       "0  ENST00000639148.1           6842  \n",
       "1  ENST00000639148.1          14787  \n",
       "2  ENST00000639148.1          19727  \n",
       "3  ENST00000639148.1          20206  \n",
       "4  ENST00000639148.1          21622  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_aggregated.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the number of peaks per gene region\n",
    "peak_count_df = df_aggregated.groupby([\"geneId\",\"annotation\"],as_index=False).agg(\n",
    "    peak_counts = ('annotation','size')\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "feature_matrix = peak_count_df.pivot_table(columns='annotation',index=['geneId'])['peak_counts']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(11382, 7)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_matrix.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th>annotation</th>\n",
       "      <th>3' UTR_H3K36me3</th>\n",
       "      <th>5' UTR_H3K36me3</th>\n",
       "      <th>Distal Intergenic_H3K36me3</th>\n",
       "      <th>Downstream_H3K36me3</th>\n",
       "      <th>Exon_H3K36me3</th>\n",
       "      <th>Intron_H3K36me3</th>\n",
       "      <th>Promoter_H3K36me3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>geneId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>5.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>12.0</td>\n",
       "      <td>17.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.0</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "annotation  3' UTR_H3K36me3  5' UTR_H3K36me3  Distal Intergenic_H3K36me3  \\\n",
       "geneId                                                                     \n",
       "1                       1.0              NaN                         NaN   \n",
       "9                       NaN              NaN                         NaN   \n",
       "14                      1.0              NaN                         NaN   \n",
       "16                      5.0              NaN                         NaN   \n",
       "19                      NaN              NaN                         NaN   \n",
       "\n",
       "annotation  Downstream_H3K36me3  Exon_H3K36me3  Intron_H3K36me3  \\\n",
       "geneId                                                            \n",
       "1                           NaN            NaN              NaN   \n",
       "9                           NaN            NaN              NaN   \n",
       "14                          1.0            NaN              NaN   \n",
       "16                          NaN            NaN             12.0   \n",
       "19                          NaN            6.0              6.0   \n",
       "\n",
       "annotation  Promoter_H3K36me3  \n",
       "geneId                         \n",
       "1                         NaN  \n",
       "9                         1.0  \n",
       "14                        NaN  \n",
       "16                       17.0  \n",
       "19                        NaN  "
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "feature_matrix.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Import RNA-seq data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq = dfs_transcriptomics['Mammary Epithelial Cell']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>length</th>\n",
       "      <th>effective_length</th>\n",
       "      <th>expected_count</th>\n",
       "      <th>TPM</th>\n",
       "      <th>FPKM</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "      <th>gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>10904</td>\n",
       "      <td>10904</td>\n",
       "      <td>93.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>BLCAP</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>22883</td>\n",
       "      <td>22883</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>CLSTN1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225</th>\n",
       "      <td>23355</td>\n",
       "      <td>23355</td>\n",
       "      <td>74.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>VPS8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>226</th>\n",
       "      <td>23461</td>\n",
       "      <td>23461</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>ABCA5</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>286</th>\n",
       "      <td>25809</td>\n",
       "      <td>25809</td>\n",
       "      <td>73.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>TTLL1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    gene_id transcript_id(s)  length  effective_length  expected_count  TPM  \\\n",
       "0     10904            10904    93.0               0.0             0.0  0.0   \n",
       "223   22883            22883    73.0               0.0             0.0  0.0   \n",
       "225   23355            23355    74.0               0.0             0.0  0.0   \n",
       "226   23461            23461    73.0               0.0             0.0  0.0   \n",
       "286   25809            25809    73.0               0.0             0.0  0.0   \n",
       "\n",
       "     FPKM  posterior_mean_count  posterior_standard_deviation_of_count  \\\n",
       "0     0.0                   0.0                                    0.0   \n",
       "223   0.0                   0.0                                    0.0   \n",
       "225   0.0                   0.0                                    0.0   \n",
       "226   0.0                   0.0                                    0.0   \n",
       "286   0.0                   0.0                                    0.0   \n",
       "\n",
       "     pme_TPM  pme_FPKM  TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "0        0.0       0.0                 0.0                 0.0   \n",
       "223      0.0       0.0                 0.0                 0.0   \n",
       "225      0.0       0.0                 0.0                 0.0   \n",
       "226      0.0       0.0                 0.0                 0.0   \n",
       "286      0.0       0.0                 0.0                 0.0   \n",
       "\n",
       "     TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "0                                      0.0                  0.0   \n",
       "223                                    0.0                  0.0   \n",
       "225                                    0.0                  0.0   \n",
       "226                                    0.0                  0.0   \n",
       "286                                    0.0                  0.0   \n",
       "\n",
       "     FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation gene_name  \n",
       "0                    0.0                                     0.0     BLCAP  \n",
       "223                  0.0                                     0.0    CLSTN1  \n",
       "225                  0.0                                     0.0      VPS8  \n",
       "226                  0.0                                     0.0     ABCA5  \n",
       "286                  0.0                                     0.0     TTLL1  "
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna_seq.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Lets harmonizome the gene identifiers in the epigenetics data and RNA-seq data to gene names "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "mapping = pd.read_table('mart_export_grch38.txt')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Gene stable ID</th>\n",
       "      <th>Gene stable ID version</th>\n",
       "      <th>Transcript stable ID</th>\n",
       "      <th>Transcript stable ID version</th>\n",
       "      <th>Gene name</th>\n",
       "      <th>NCBI gene (formerly Entrezgene) ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ENSG00000210049</td>\n",
       "      <td>ENSG00000210049.1</td>\n",
       "      <td>ENST00000387314</td>\n",
       "      <td>ENST00000387314.1</td>\n",
       "      <td>MT-TF</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>ENSG00000211459</td>\n",
       "      <td>ENSG00000211459.2</td>\n",
       "      <td>ENST00000389680</td>\n",
       "      <td>ENST00000389680.2</td>\n",
       "      <td>MT-RNR1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>ENSG00000210077</td>\n",
       "      <td>ENSG00000210077.1</td>\n",
       "      <td>ENST00000387342</td>\n",
       "      <td>ENST00000387342.1</td>\n",
       "      <td>MT-TV</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>ENSG00000210082</td>\n",
       "      <td>ENSG00000210082.2</td>\n",
       "      <td>ENST00000387347</td>\n",
       "      <td>ENST00000387347.2</td>\n",
       "      <td>MT-RNR2</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ENSG00000209082</td>\n",
       "      <td>ENSG00000209082.1</td>\n",
       "      <td>ENST00000386347</td>\n",
       "      <td>ENST00000386347.1</td>\n",
       "      <td>MT-TL1</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    Gene stable ID Gene stable ID version Transcript stable ID  \\\n",
       "0  ENSG00000210049      ENSG00000210049.1      ENST00000387314   \n",
       "1  ENSG00000211459      ENSG00000211459.2      ENST00000389680   \n",
       "2  ENSG00000210077      ENSG00000210077.1      ENST00000387342   \n",
       "3  ENSG00000210082      ENSG00000210082.2      ENST00000387347   \n",
       "4  ENSG00000209082      ENSG00000209082.1      ENST00000386347   \n",
       "\n",
       "  Transcript stable ID version Gene name  NCBI gene (formerly Entrezgene) ID  \n",
       "0            ENST00000387314.1     MT-TF                                 NaN  \n",
       "1            ENST00000389680.2   MT-RNR1                                 NaN  \n",
       "2            ENST00000387342.1     MT-TV                                 NaN  \n",
       "3            ENST00000387347.2   MT-RNR2                                 NaN  \n",
       "4            ENST00000386347.1    MT-TL1                                 NaN  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "mapping.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/p4/gdh28gk568sbf143d4p59vdw0000gq/T/ipykernel_13939/757272775.py:6: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'] = ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'].astype(int).astype(str)\n"
     ]
    }
   ],
   "source": [
    "ensembl_gene_id_map = mapping.set_index('Gene stable ID version').to_dict()['Gene name']\n",
    "ensembl_transcript_id_map = mapping.set_index('Transcript stable ID version').to_dict()['Gene name']\n",
    "\n",
    "# NCBI column has NaNs which turns the integers into floats, so drop NaNs, change floats --> int --> str, and create dict\n",
    "ncbi_gene_id_map = mapping.dropna(subset='NCBI gene (formerly Entrezgene) ID')\n",
    "ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'] = ncbi_gene_id_map['NCBI gene (formerly Entrezgene) ID'].astype(int).astype(str)\n",
    "ncbi_gene_id_map = ncbi_gene_id_map.set_index('NCBI gene (formerly Entrezgene) ID').to_dict()['Gene name']\n",
    "\n",
    "# Combine the transcript IDs, gene IDs, and NCBI IDs dictionaries for all possible mappings\n",
    "mapping_dict = ensembl_transcript_id_map | ensembl_gene_id_map | ncbi_gene_id_map "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq['gene_name'] = rna_seq['gene_id'].map(mapping_dict)\n",
    "feature_matrix['gene_name'] = feature_matrix.index.astype(str).map(mapping_dict)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "rna_seq.dropna(subset='gene_name', inplace = True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The below seems like a lot of genes aren't mapped between the epigenetics and RNA-seq datasets... this limits our training data quite a bit"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>3' UTR_H3K36me3</th>\n",
       "      <th>5' UTR_H3K36me3</th>\n",
       "      <th>Distal Intergenic_H3K36me3</th>\n",
       "      <th>Downstream_H3K36me3</th>\n",
       "      <th>Exon_H3K36me3</th>\n",
       "      <th>Intron_H3K36me3</th>\n",
       "      <th>Promoter_H3K36me3</th>\n",
       "      <th>gene_name</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>transcript_id(s)</th>\n",
       "      <th>...</th>\n",
       "      <th>posterior_mean_count</th>\n",
       "      <th>posterior_standard_deviation_of_count</th>\n",
       "      <th>pme_TPM</th>\n",
       "      <th>pme_FPKM</th>\n",
       "      <th>TPM_ci_lower_bound</th>\n",
       "      <th>TPM_ci_upper_bound</th>\n",
       "      <th>TPM_coefficient_of_quartile_variation</th>\n",
       "      <th>FPKM_ci_lower_bound</th>\n",
       "      <th>FPKM_ci_upper_bound</th>\n",
       "      <th>FPKM_coefficient_of_quartile_variation</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>12.0</td>\n",
       "      <td>13.0</td>\n",
       "      <td>JAG1</td>\n",
       "      <td>ENSG00000101384.12</td>\n",
       "      <td>ENST00000254958.10,ENST00000423891.6,ENST00000...</td>\n",
       "      <td>...</td>\n",
       "      <td>11444.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>99.75</td>\n",
       "      <td>105.99</td>\n",
       "      <td>96.428500</td>\n",
       "      <td>103.203000</td>\n",
       "      <td>0.011759</td>\n",
       "      <td>102.481000</td>\n",
       "      <td>109.678000</td>\n",
       "      <td>0.011786</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>3.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>7.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>2.0</td>\n",
       "      <td>ALCAM</td>\n",
       "      <td>ENSG00000170017.12</td>\n",
       "      <td>ENST00000306107.9,ENST00000460954.1,ENST000004...</td>\n",
       "      <td>...</td>\n",
       "      <td>3032.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>27.99</td>\n",
       "      <td>29.74</td>\n",
       "      <td>26.401500</td>\n",
       "      <td>29.539100</td>\n",
       "      <td>0.019641</td>\n",
       "      <td>28.076200</td>\n",
       "      <td>31.414700</td>\n",
       "      <td>0.019630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>C1QBP</td>\n",
       "      <td>ENSG00000108561.8</td>\n",
       "      <td>ENST00000225698.8,ENST00000570805.1,ENST000005...</td>\n",
       "      <td>...</td>\n",
       "      <td>2415.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>90.66</td>\n",
       "      <td>96.33</td>\n",
       "      <td>86.624700</td>\n",
       "      <td>94.748200</td>\n",
       "      <td>0.015599</td>\n",
       "      <td>92.040600</td>\n",
       "      <td>100.678000</td>\n",
       "      <td>0.015580</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>GAS8-AS1</td>\n",
       "      <td>ENSG00000221819.6</td>\n",
       "      <td>ENST00000408886.4,ENST00000623094.2</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.15</td>\n",
       "      <td>0.16</td>\n",
       "      <td>0.003490</td>\n",
       "      <td>0.390104</td>\n",
       "      <td>0.499912</td>\n",
       "      <td>0.003707</td>\n",
       "      <td>0.414650</td>\n",
       "      <td>0.499905</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>S100G</td>\n",
       "      <td>ENSG00000169906.5</td>\n",
       "      <td>ENST00000380200.3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.317267</td>\n",
       "      <td>0.657416</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.337123</td>\n",
       "      <td>0.657527</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1132</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>ZNF710-AS1</td>\n",
       "      <td>ENSG00000259291.2</td>\n",
       "      <td>ENST00000558334.1,ENST00000620791.1</td>\n",
       "      <td>...</td>\n",
       "      <td>12.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.36</td>\n",
       "      <td>0.38</td>\n",
       "      <td>0.180287</td>\n",
       "      <td>0.560744</td>\n",
       "      <td>0.189055</td>\n",
       "      <td>0.191685</td>\n",
       "      <td>0.596078</td>\n",
       "      <td>0.188964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1133</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>LAMC1-AS1</td>\n",
       "      <td>ENSG00000224468.3</td>\n",
       "      <td>ENST00000457852.3</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.10</td>\n",
       "      <td>0.11</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>0.295451</td>\n",
       "      <td>0.657439</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.313975</td>\n",
       "      <td>0.657322</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1134</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>2.0</td>\n",
       "      <td>MVP-DT</td>\n",
       "      <td>ENSG00000238045.9</td>\n",
       "      <td>ENST00000563806.1,ENST00000564980.1,ENST000005...</td>\n",
       "      <td>...</td>\n",
       "      <td>22.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>1.14</td>\n",
       "      <td>1.22</td>\n",
       "      <td>0.616980</td>\n",
       "      <td>1.708690</td>\n",
       "      <td>0.167849</td>\n",
       "      <td>0.653801</td>\n",
       "      <td>1.814020</td>\n",
       "      <td>0.167858</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1135</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>4.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>3.0</td>\n",
       "      <td>ABCF2-H2BK1</td>\n",
       "      <td>ENSG00000285292.1</td>\n",
       "      <td>ENST00000222388.6</td>\n",
       "      <td>...</td>\n",
       "      <td>297.64</td>\n",
       "      <td>49.22</td>\n",
       "      <td>5.39</td>\n",
       "      <td>5.72</td>\n",
       "      <td>3.601470</td>\n",
       "      <td>7.188580</td>\n",
       "      <td>0.122905</td>\n",
       "      <td>3.828690</td>\n",
       "      <td>7.642080</td>\n",
       "      <td>0.122929</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1136</th>\n",
       "      <td>1.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>3.0</td>\n",
       "      <td>8.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>RCC2-AS1</td>\n",
       "      <td>ENSG00000227751.1</td>\n",
       "      <td>ENST00000439577.1</td>\n",
       "      <td>...</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.00</td>\n",
       "      <td>0.18</td>\n",
       "      <td>0.19</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.536593</td>\n",
       "      <td>0.657900</td>\n",
       "      <td>0.000004</td>\n",
       "      <td>0.570103</td>\n",
       "      <td>0.657952</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1137 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "      3' UTR_H3K36me3  5' UTR_H3K36me3  Distal Intergenic_H3K36me3  \\\n",
       "0                 2.0              NaN                         NaN   \n",
       "1                 3.0              NaN                         NaN   \n",
       "2                 8.0              NaN                         NaN   \n",
       "3                 1.0              NaN                         NaN   \n",
       "4                 NaN              NaN                         NaN   \n",
       "...               ...              ...                         ...   \n",
       "1132              NaN              NaN                         NaN   \n",
       "1133              1.0              NaN                         NaN   \n",
       "1134              NaN              NaN                         NaN   \n",
       "1135              NaN              NaN                         NaN   \n",
       "1136              1.0              NaN                         NaN   \n",
       "\n",
       "      Downstream_H3K36me3  Exon_H3K36me3  Intron_H3K36me3  Promoter_H3K36me3  \\\n",
       "0                     NaN            4.0             12.0               13.0   \n",
       "1                     NaN            7.0              3.0                2.0   \n",
       "2                     NaN            1.0              4.0                3.0   \n",
       "3                     NaN            NaN              NaN                NaN   \n",
       "4                     NaN            NaN              2.0                NaN   \n",
       "...                   ...            ...              ...                ...   \n",
       "1132                  NaN            NaN              1.0                NaN   \n",
       "1133                  NaN            1.0              NaN                2.0   \n",
       "1134                  NaN            NaN              NaN                2.0   \n",
       "1135                  NaN            4.0              3.0                3.0   \n",
       "1136                  NaN            3.0              8.0                NaN   \n",
       "\n",
       "        gene_name             gene_id  \\\n",
       "0            JAG1  ENSG00000101384.12   \n",
       "1           ALCAM  ENSG00000170017.12   \n",
       "2           C1QBP   ENSG00000108561.8   \n",
       "3        GAS8-AS1   ENSG00000221819.6   \n",
       "4           S100G   ENSG00000169906.5   \n",
       "...           ...                 ...   \n",
       "1132   ZNF710-AS1   ENSG00000259291.2   \n",
       "1133    LAMC1-AS1   ENSG00000224468.3   \n",
       "1134       MVP-DT   ENSG00000238045.9   \n",
       "1135  ABCF2-H2BK1   ENSG00000285292.1   \n",
       "1136     RCC2-AS1   ENSG00000227751.1   \n",
       "\n",
       "                                       transcript_id(s)  ...  \\\n",
       "0     ENST00000254958.10,ENST00000423891.6,ENST00000...  ...   \n",
       "1     ENST00000306107.9,ENST00000460954.1,ENST000004...  ...   \n",
       "2     ENST00000225698.8,ENST00000570805.1,ENST000005...  ...   \n",
       "3                   ENST00000408886.4,ENST00000623094.2  ...   \n",
       "4                                     ENST00000380200.3  ...   \n",
       "...                                                 ...  ...   \n",
       "1132                ENST00000558334.1,ENST00000620791.1  ...   \n",
       "1133                                  ENST00000457852.3  ...   \n",
       "1134  ENST00000563806.1,ENST00000564980.1,ENST000005...  ...   \n",
       "1135                                  ENST00000222388.6  ...   \n",
       "1136                                  ENST00000439577.1  ...   \n",
       "\n",
       "      posterior_mean_count  posterior_standard_deviation_of_count  pme_TPM  \\\n",
       "0                 11444.00                                   0.00    99.75   \n",
       "1                  3032.00                                   0.00    27.99   \n",
       "2                  2415.00                                   0.00    90.66   \n",
       "3                     0.00                                   0.00     0.15   \n",
       "4                     0.00                                   0.00     0.11   \n",
       "...                    ...                                    ...      ...   \n",
       "1132                 12.00                                   0.00     0.36   \n",
       "1133                  0.00                                   0.00     0.10   \n",
       "1134                 22.00                                   0.00     1.14   \n",
       "1135                297.64                                  49.22     5.39   \n",
       "1136                  0.00                                   0.00     0.18   \n",
       "\n",
       "      pme_FPKM  TPM_ci_lower_bound  TPM_ci_upper_bound  \\\n",
       "0       105.99           96.428500          103.203000   \n",
       "1        29.74           26.401500           29.539100   \n",
       "2        96.33           86.624700           94.748200   \n",
       "3         0.16            0.003490            0.390104   \n",
       "4         0.11            0.000003            0.317267   \n",
       "...        ...                 ...                 ...   \n",
       "1132      0.38            0.180287            0.560744   \n",
       "1133      0.11            0.000002            0.295451   \n",
       "1134      1.22            0.616980            1.708690   \n",
       "1135      5.72            3.601470            7.188580   \n",
       "1136      0.19            0.000007            0.536593   \n",
       "\n",
       "      TPM_coefficient_of_quartile_variation  FPKM_ci_lower_bound  \\\n",
       "0                                  0.011759           102.481000   \n",
       "1                                  0.019641            28.076200   \n",
       "2                                  0.015599            92.040600   \n",
       "3                                  0.499912             0.003707   \n",
       "4                                  0.657416             0.000003   \n",
       "...                                     ...                  ...   \n",
       "1132                               0.189055             0.191685   \n",
       "1133                               0.657439             0.000003   \n",
       "1134                               0.167849             0.653801   \n",
       "1135                               0.122905             3.828690   \n",
       "1136                               0.657900             0.000004   \n",
       "\n",
       "      FPKM_ci_upper_bound  FPKM_coefficient_of_quartile_variation  \n",
       "0              109.678000                                0.011786  \n",
       "1               31.414700                                0.019630  \n",
       "2              100.678000                                0.015580  \n",
       "3                0.414650                                0.499905  \n",
       "4                0.337123                                0.657527  \n",
       "...                   ...                                     ...  \n",
       "1132             0.596078                                0.188964  \n",
       "1133             0.313975                                0.657322  \n",
       "1134             1.814020                                0.167858  \n",
       "1135             7.642080                                0.122929  \n",
       "1136             0.570103                                0.657952  \n",
       "\n",
       "[1137 rows x 25 columns]"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.merge(feature_matrix,rna_seq,on='gene_name')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "biohack",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
